src/table/use-sticky-columns.tsx

```diff
@@ -0,0 +1,228 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { useLayoutEffect, useState, createRef, useEffect, useCallback } from 'react';
+import { useVisualRefresh } from '../internal/hooks/use-visual-mode';
+import { useIntersectionObserver } from '../internal/hooks/use-intersection-observer';
+import { TableProps } from './interfaces';
+interface CellWidths {
+  start: number[];
+  end: number[];
+}
+interface StickyStyles {
+  left?: string;
+  right?: string;
+  paddingLeft?: string;
+}
+
+interface StickyColumnParams {
+  containerWidth: number | null;
+  hasSelection: boolean;
+  isWrapperScrollable: boolean;
+  stickyColumns?: TableProps.StickyColumns;
+  visibleColumnsLength: number;
+  tableRefObject: React.RefObject<HTMLTableElement>;
+  wrapperRefObject: React.RefObject<HTMLDivElement>;
+}
+
+export interface GetStickyColumn {
+  isStickyLeft: boolean;
+  isStickyRight: boolean;
+  isLastStickyLeft: boolean;
+  isLastStickyRight: boolean;
+  stickyStyles: StickyStyles;
+}
+
+const getPadding = (element: HTMLElement | null, side: 'Left' | 'Right') =>
+  element ? Number(getComputedStyle(element)[`padding${side}`].slice(0, -2)) : 0;
+
+// We allow the table to have a minimum of 148px of available space besides the sum of the widths of the sticky columns
+const MINIMUM_SCROLLABLE_SPACE = 148;
+
+export const useStickyState = ({
+  tableRefObject,
+  wrapperRefObject,
+}: {
+  tableRefObject: React.RefObject<HTMLTableElement>;
+  wrapperRefObject: React.RefObject<HTMLDivElement>;
+}) => {
+  const [isStuckToTheLeft, setIsStuckToTheLeft] = useState(false);
+  const [isStuckToTheRight, setIsStuckToTheRight] = useState(false);
+
+  const intersectionObserverOptions = { threshold: [0, 1], rootMargin: '-2px' };
```

Comment:
I've seen this 2px offset in CSS for right sentinel. Would it make sense to document it explicitly or use constants?